### PR TITLE
Map-style Dataset to IterableDataset

### PR DESCRIPTION
**Pull request by lhoestq**
Added `ds.to_iterable()` to get an iterable dataset from a map-style arrow dataset.

It also has a `num_shards` argument to split the dataset before converting to an iterable dataset. Sharding is important to enable efficient shuffling and parallel loading of iterable datasets.

TODO:
- [ ] tests
- [ ] docs

Fix https://github.com/huggingface/datasets/issues/5265

### CODE REVIEW

File: src/datasets/builder.py [hunk 1]
```diff
@@ -493,7 +493,7 @@ def _create_builder_config(
         )
         is_custom = (config_id not in self.builder_configs) and config_id != "default"
         if is_custom:
-            logger.warning(f"Using custom data configuration {config_id}")
+            logger.info(f"Using custom data configuration {config_id}")
```
Comment (lhoestq) [inline src/datasets/builder.py:496]: I did this because I think it's not relevant anymore, and because I find it confusing to show this when calling `IterableDataset.from_generator`

File: docs/source/dataset_vs_iterable_dataset.mdx [hunk 2]
```diff
@@ -0,0 +1,217 @@
+# Choose between Dataset and IterableDataset
+
+## Downloading and streaming
+
+When you have a regular "map-style" [`Dataset`], you can access it using `my_dataset[0]`: we have what we call "random access" to the rows.
```
Comment (stevhliu) [inline docs/source/dataset_vs_iterable_dataset.mdx:5]: What is meant by a “map-style” `Dataset`? If I understand correctly, this is just a regular `Dataset`. So it might be easier for users to understand if we don’t use this specific term and just use `Dataset` or if we define what we mean by “map-style” (unless this is commonly known jargon, in which case ignore this haha).

File: docs/source/dataset_vs_iterable_dataset.mdx [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+## Loading local files entirely and progressively
+
+It is possible to convert local or remote data files to an Arrow [`Dataset`] using [`load_dataset`]:
+```python
+data_files = {"train": ["path/to/data.csv"]}
+my_dataset = load_dataset("csv", data_files=data_files, split="train")
+print(my_dataset[0])
+```
+
+To not have to wait for the conversion to Arrow, you can define an iterable dataset by streaming from your local files.
```
Comment (stevhliu) [inline docs/source/dataset_vs_iterable_dataset.mdx:68]: What is the benefit of not converting Dataset to Arrow (obvs it’s faster, but it’d be good to mention this explicitly for the user)?

**Review by stevhliu — COMMENTED** [review]
This is so cool, I learned a lot reading this. I'm sure it'll be super valuable and welcomed by the community! :smile: 

I think this would be more of a Conceptual Guide doc since this is more explanatory and compares the differences between a `Dataset` and an `IterableDataset`. It’s not necessarily a how-to for how to do something, but it discusses and explains the two types of datasets. There are definitely places in the docs where we can add a nice <Tip> and link to this doc though to build up the user's understanding of this topic. For example, in the _Know your dataset_ [tutorial](https://huggingface.co/docs/datasets/access), we only introduce the regular `Dataset` object and not the `IterableDataset`. We can add a section there for `IterableDataset` and then link to this doc that explains the difference between the two :slightly_smiling_face:

File: docs/source/dataset_vs_iterable_dataset.mdx (hunk 2 above)
Comment (lhoestq) [reply docs/source/dataset_vs_iterable_dataset.mdx:5]: Yes it refers to datasets with random access, i.e. that allows you to do `my_dataset[0]`. I'll define it properly

File: docs/source/dataset_vs_iterable_dataset.mdx (hunk 3 above)
Comment (lhoestq) [reply docs/source/dataset_vs_iterable_dataset.mdx:68]: Faster + save disk space + you can modify your original data and re-instantiate the dataset without having to reconvert the original data. I'll mention this !

**Comment by lhoestq** [comment]
> I think this would be more of a Conceptual Guide doc since this is more explanatory and compares the differences between a Dataset and an IterableDataset

sounds good to me !

> There are definitely places in the docs where we can add a nice and link to this doc though to build up the user's understanding of this topic. For example, in the Know your dataset [tutorial](https://huggingface.co/docs/datasets/access), we only introduce the regular Dataset object and not the IterableDataset. We can add a section there for IterableDataset and then link to this doc that explains the difference between the two 🙂

good idea, thanks :)

**Comment by stevhliu** [comment]
I'll open a PR to add a section on `IterableDataset`'s in the tutorial, and once you're done editing this doc I can give it a final polish! 😄

**Comment by lhoestq** [comment]
I moved the doc page to conceptual guides and took your suggestions into account :)

I think this is ready for final review now

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 4]
```diff
@@ -0,0 +1,220 @@
+# Differences between Dataset and IterableDataset
+
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:2]: Maybe just add a sentence or two here that introduces the topic and scope of the doc. Something like:

There are two types of dataset objects, a `Dataset` and an `IterableDataset`. Whichever type of dataset you choose to use or create depends on the size of the dataset. In general, an `IterableDataset` is ideal for big datasets (think hundreds of GBs!) due to its lazy behavior and speed advantages, while a `Dataset` is great for everything else. This page will compare the differences between a `Dataset` and an `IterableDataset` to help you pick the right dataset object for you.

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+## Downloading and streaming
+
+When you have a regular [`Dataset`], you can access it using `my_dataset[0]`: we have what we call "random access" to the rows.
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:5]:
```suggestion
When you have a regular [`Dataset`], you can access it using `my_dataset[0]`. This provides "random access" to the rows.
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+from datasets import load_dataset
+
+imagenet = load_dataset("imagenet-1k", split="train")  # downloads the full dataset
+print(imagenet[0])
+```
+
+But one caveat is that you must have all the dataset stored on your disk or in memory, which makes it not always suitable for datasets bigger than disk.
+Because it can become inconvenient for big datasets, there exists another type of dataset: [`IterableDataset`].
+When you have an iterable dataset, you can access it using a `for` loop to load the data progressively as you iterate over the dataset.
+This way, only a small fraction of examples is loaded in memory, and you don't write anything on disk.
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:19]:
```suggestion
But one caveat is that you must have the entire dataset stored on your disk or in memory, which blocks you from accessing datasets bigger than the disk.
Because it can become inconvenient for big datasets, there exists another type of dataset, the [`IterableDataset`].
When you have an `IterableDataset`, you can access it using a `for` loop to load the data progressively as you iterate over the dataset.
This way, only a small fraction of examples is loaded in memory, and you don't write anything on disk.
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+For example, you can stream the ImageNet-1k dataset without downloading it completely:
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:21]: I think it'd be good to emphasize here that you don't have to wait 🐢 

```suggestion
For example, you can stream the ImageNet-1k dataset without waiting for it to download completely:
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+
+imagenet = load_dataset("imagenet-1k", split="train", streaming=True)  # will start loading the data when iterated over
+for example in imagenet:
+    print(example)
+    break
+```
+
+Dataset streaming can be used to read online data without writing any file on disk.
+For example you can stream datasets made out of multiple hundreds of gigabytes like [C4](https://huggingface.co/datasets/c4), [OSCAR](https://huggingface.co/datasets/oscar) or [LAION-2B](https://huggingface.co/datasets/laion/laion2B-en).
+Learn more about dataset streaming in the [Dataset Streaming Guide](./stream).
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:34]:
```suggestion
Streaming can read online data without writing any file to disk.
For example, you can stream datasets made out of multiple shards, each of which is hundreds of gigabytes like [C4](https://huggingface.co/datasets/c4), [OSCAR](https://huggingface.co/datasets/oscar) or [LAION-2B](https://huggingface.co/datasets/laion/laion2B-en).
Learn more about how to stream a dataset in the [Dataset Streaming Guide](./stream).
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+
+This is not the only difference.
+This "lazy" behavior of iterable datasets is also present when it comes to dataset creation and processing.
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:37]:
```suggestion
This is not the only difference though, because the "lazy" behavior of an `IterableDataset` is also present when it comes to dataset creation and processing.
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+
+## Creating map-style datasets and iterable datasets
+
+You can create a [`Dataset`] using lists or dictionaries, and the data is entirely converted to Arrow so you can access the rows easily
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:41]:
```suggestion
You can create a [`Dataset`] using lists or dictionaries, and the data is entirely converted to Arrow so you can easily access any row:
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+```python
+my_dataset = Dataset.from_dict({"col_1": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]})
+print(my_dataset[0])
+```
+
+To create an iterable dataset on the other hand, you must provide a "lazy" way to load the data.
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:47]:
```suggestion
To create an `IterableDataset` on the other hand, you must provide a "lazy" way to load the data.
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+In Python, we generally use generator functions. These functions `yield` one example at a time:
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:48]:
```suggestion
In Python, we generally use generator functions. These functions `yield` one example at a time, which means you can't access a row by slicing it like a regular `Dataset`:
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+## Loading local files entirely and progressively
+
+It is possible to convert local or remote data files to an Arrow [`Dataset`] using [`load_dataset`]:
+```python
+data_files = {"train": ["path/to/data.csv"]}
+my_dataset = load_dataset("csv", data_files=data_files, split="train")
+print(my_dataset[0])
+```
+
+However here it requires a conversion step from CSV to Arrow format, which can take time and disk space if your dataset is big.
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:69]:
```suggestion
However, this requires a conversion step from CSV to Arrow format, which takes time and disk space if your dataset is big.
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+
+To save disk space ans skip the conversion step, you can define an iterable dataset by streaming from the local files directly.
+This way the data is read progessively from the local files as you iterate over the dataset:
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:72]:
```suggestion
To save disk space and skip the conversion step, you can define an `IterableDataset` by streaming from the local files directly.
This way, the data is read progressively from the local files as you iterate over the dataset:
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+
+```python
+data_files = {"train": ["path/to/data.csv"]}
+my_iterable_dataset = load_dataset("csv", data_files=data_files, split="train", streaming=True)
+for example in my_iterable_dataset:  # this reads the CSV file progressively as you iterate over the dataset
+    print(example)
+    break
+```
+
+Many files formats are supported like CSV, JSONL, Parquet, images, audio files etc.
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:82]:
```suggestion
Many file formats are supported, like CSV, JSONL, and Parquet, as well as image and audio files.
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+You can find more information in the corresponding guides for loading [tabular](./tabular_load), [text](./nlp_load), [vision]()./image_load) and [audio](./audio_load]) datasets.
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:83]:
```suggestion
You can find more information in the corresponding guides for loading [tabular](./tabular_load), [text](./nlp_load), [vision](./image_load), and [audio](./audio_load]) datasets.
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+
+## Eager data processing and lazy data processing
+
+When you process a [`Dataset`] object using [`Dataset.map`] for example, the data processing happens immediatly when you call [`Dataset.map`] and returns the processed dataset.
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:87]:
```suggestion
When you process a [`Dataset`] object using [`Dataset.map`], the entire dataset is processed immediately and returned.
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+This is similar to how `pandas` works for example.
+
+```python
+my_dataset = my_dataset.map(process_fn)  # process_fn is applied on all the examples of the dataset
+print(my_dataset[0])
+```
+
+On the other hand, due to the "lazy" nature of iterable datasets, calling [`IterableDataset.map`] does not iterate over the full dataset to apply your `map` function.
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:95]:
```suggestion
On the other hand, due to the "lazy" nature of an `IterableDataset`, calling [`IterableDataset.map`] does not apply your `map` function over the full dataset.
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+Instead, your `map` function is applied on-the-fly when you iterate over the iterable dataset.
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:96]:
```suggestion
Instead, your `map` function is applied on-the-fly.
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+for example in my_iterable_dataset:  
+    print(example)
+    break
+```
+
+## Exact and fast approximate shuffling
+
+When you shuffle a [`Dataset`] using [`Dataset.shuffle`], you apply an exact shuffling of the dataset.
+It works by taking a list of indices `[0, 1, 2, ... len(my_dataset) - 1]` and shuffling this list.
+Then, accessing `my_dataset[0]` accesses the row and the index defined by the first element of the indices mapping that has been shuffled:
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:115]:
```suggestion
Then, accessing `my_dataset[0]` returns the row and index defined by the first element of the indices mapping that has been shuffled:
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+```python
+my_dataset = my_dataset.shuffle(seed=42)
+print(my_dataset[0])
+```
+
+Since we don't have random access to the rows in the case of iterable datasets, we can't use a shuffled list of indices and access a row at an arbitrary position.
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:121]:
```suggestion
Since we don't have random access to the rows in the case of an `IterableDataset`, we can't use a shuffled list of indices and access a row at an arbitrary position.
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 22]
```diff
@@ -0,0 +1,220 @@
...
+It uses a shuffle buffer to sample random examples iteratively from the dataset.
+Since the dataset is still read iteratively, it provides excellent speed performance:
+```python
+my_iterable_dataset = my_iterable_dataset.shuffle(seed=42, buffer_size=100)
+for example in my_iterable_dataset:
+    print(example)
+    break
+```
+
+Since using a shuffle buffer is not enough to provide a satisfactory shuffling for machine learning model training, [`IterableDataset.shuffle`] also shuffles the dataset shards if your dataset is made of multiple files or sources:
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:133]:
```suggestion
But using a shuffle buffer is not enough to provide a satisfactory shuffling for machine learning model training. So [`IterableDataset.shuffle`] also shuffles the dataset shards if your dataset is made of multiple files or sources:
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 23]
```diff
@@ -0,0 +1,220 @@
...
+my_iterable_dataset = IterableDataset.from_generator(my_generator, gen_kwargs=gen_kwargs)
+my_iterable_dataset.n_shards  # 1024
+```
+
+## Speed differences
+
+Regular [`Dataset`] objects are based on Arrow which provides fast random access to the rows.
+Thanks to memory mapping and the fact that Arrow is an in-memory format, reading data from disk doesn't do expensive system calls and deserialization.
+
+It also provides even faster data loading when iterating using a `for` loop by iterating on contiguous Arrow record batches.
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:161]: I'd combine this sentence with the above, looks a little strange on its own :P

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 24]
```diff
@@ -0,0 +1,220 @@
...
+
+However as soon as your [`Dataset`] has an indices mapping (via [`Dataset.shuffle`] for example), the speed can become 10x slower.
+This is because there is an extra step to get the row index to read using the indices mapping, and most importantly because you no longer read contiguous chunks of data.
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:164]:
```suggestion
This is because there is an extra step to get the row index to read using the indices mapping, and most importantly, you aren't reading contiguous chunks of data anymore.
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 25]
```diff
@@ -0,0 +1,220 @@
...
+To get the optimal speed back, you'd need to rewrite the entire dataset on your disk again using `Dataset.flatten_indices` which removes the indices mapping.
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:165]:
```suggestion
To restore the speed, you'd need to rewrite the entire dataset on your disk again using [`Dataset.flatten_indices`], which removes the indices mapping.
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 26]
```diff
@@ -0,0 +1,220 @@
...
+```python
+my_dataset[0]  # fast
+my_dataset = my_dataset.shuffle(seed=42)
+my_dataset[0]  # up to 10x slower
+my_dataset = my_dataset.flatten_indices()  # rewrite the shuffled dataset on disk as contiguous chunks of data
+my_dataset[0]  # fast again
+```
+
+
+In this case is it recommended to switch to an [`IterableDataset`] and leverage its fast approximate shuffling method [`IterableDataset.shuffle`].
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:177]:
```suggestion
In this case, we recommend switching to an [`IterableDataset`] and leveraging its fast approximate shuffling method [`IterableDataset.shuffle`].
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 27]
```diff
@@ -0,0 +1,220 @@
...
+
+shuffled_iterable_dataset = my_iterable_dataset.shuffle(seed=1337, buffer_size=100)  # reshuffling using another seed is instantaneous
+
+for example in enumerate(shuffled_iterable_dataset):  # still as fast as before
+    pass
+```
+
+Note if you use your dataset on multiple epochs:
+
+The effective seed to shuffle the shards order and for the shuffle buffer is `seed + epoch`.
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:198]:
```suggestion
If you're using your dataset on multiple epochs, the effective seed to shuffle the shards order in the shuffle buffer is `seed + epoch`.
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 28]
```diff
@@ -0,0 +1,220 @@
...
+```python
+for epoch in range(n_epochs):
+    my_iterable_dataset.set_epoch(epoch)
+    for example in my_iterable_dataset:  # fast + reshuffled at each epoch using `effective_seed = seed + epoch`
+        pass
+```
+
+## Switch from map-style to iterable
+
+If you want to benefit from the "lazy" behavior of iterable datasets or their speed advantages, you can switch your map-style [`Dataset`] to an [`IterableDataset`]:
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:209]:
```suggestion
If you want to benefit from the "lazy" behavior of an  [`IterableDataset`] or their speed advantages, you can switch your map-style [`Dataset`] to an [`IterableDataset`]:
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 29]
```diff
@@ -0,0 +1,220 @@
...
+```python
+my_iterable_dataset = my_dataset.to_iterable()
+```
+
+If you want to shuffle your dataset or [use it with a PyTorch DataLoader](./use_with_pytorch#stream-data), it is recommended to generate a sharded iterable dataset:
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:214]:
```suggestion
If you want to shuffle your dataset or [use it with a PyTorch DataLoader](./use_with_pytorch#stream-data), we recommend generating a shared [`IterableDataset`]:
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 30]
```diff
@@ -0,0 +1,220 @@
...
+```python
+my_iterable_dataset = my_dataset.to_iterable(num_shards=1024)
+my_iterable_dataset.n_shards  # 1024
+```
+
+Feel free to open a discussion on the 🤗 Datasets [forum](https://discuss.huggingface.co/c/datasets/10) if you have questions !
```
Comment (stevhliu) [inline docs/source/about_mapstyle_vs_iterable.mdx:220]:
```suggestion
Feel free to open a discussion on the 🤗 Datasets [forum](https://discuss.huggingface.co/c/datasets/10) if you have questions!
```

File: src/datasets/arrow_dataset.py [hunk 31]
```diff
@@ -4643,6 +4644,120 @@ def extra_nbytes_visitor(array, feature):
...
+
+        Contrary to map-style datasets, iterable datasets are lazy and can only be iterated over (e.g. using a for loop).
+        Since they are read sequentially in training loops, iterable datasets are much faster than map-style datasets.
+        All the transformations applied to iterable datasets like filtering or processing are done on-the-fly when you start iterating over the dataset.
+
+        Still, it is possible to shuffle an iterable dataset using [`datasets.IterableDataset.shuffle`].
+        This is a fast approximate shuffling that works best if you have multiple shards and if you specify a buffer size that is big enough.
+
+        To get the best speed performance, make sure your dataset doesn't have an indices mapping.
+        Indeed if this is the case, the data are not read contiguously, which can be slow sometimes.
```
Comment (stevhliu) [inline src/datasets/arrow_dataset.py:4664]:
```suggestion
        If this is the case, the data are not read contiguously, which can be slow sometimes.
```

File: src/datasets/arrow_dataset.py [hunk 32]
```diff
@@ -4643,6 +4644,120 @@ def extra_nbytes_visitor(array, feature):
...
+        Args:
+            num_shards (`int`, default to `1`):
+                Number of shards to define when instantiating the iterable dataset. This is especially useful for big datasets to be able to shuffle properly,
+                and also to enable fast parallel loading using a PyTorch DataLoader or in distributed setups for example.
+                Shards are defined using [`datasets.Dataset.shard`]: it simply slices the data without writing anything on disk.
+
+        Returns:
+            [`datasets.IterableDataset`]
+
+        Example:
```
Comment (stevhliu) [inline src/datasets/arrow_dataset.py:4676]: Love all the example usages here! 😍

**Review by stevhliu — APPROVED** [review]
Awesome doc, thanks for sharing all this infos!

File: docs/source/about_mapstyle_vs_iterable.mdx (hunk 30 above)
Comment (mariosasko) [reply docs/source/about_mapstyle_vs_iterable.mdx:220]: I would remove this sentence altogether. Two existing links in our docs are more than enough  :).

**Review by mariosasko — COMMENTED** [review]
Nice!

The code looks good.

Regarding the docs, I think it would be better to add this info as notes/tips/sections to the existing docs (Process/Stream; e.g. a tip under `Dataset.shuffle` that explains how to make this operation more performant by using `to_iterable` + `shuffle`, etc.) rather than introducing a new doc page.

File: docs/source/use_with_pytorch.mdx [hunk 33]
```diff
@@ -200,7 +200,12 @@ You must use a `BatchSampler` if you want the transform to be given full batches
 
 ### Stream data
 
-Loading a dataset in streaming mode allows one to iterate over the dataset without downloading it on disk.
+
```
Comment (lhoestq) [inline docs/source/use_with_pytorch.mdx:203]:
```suggestion
```

**Comment by lhoestq** [comment]
I took your comments into account :)

> Regarding the docs, I think it would be better to add this info as notes/tips/sections to the existing docs (Process/Stream; e.g. a tip under Dataset.shuffle that explains how to make this operation more performant by using to_iterable + shuffle, etc.) rather than introducing a new doc page.

I added a paragraph in the Dataset.shuffle docstring, and a note in the Process doc page

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 34]
```diff
@@ -0,0 +1,219 @@
...
+imagenet = load_dataset("imagenet-1k", split="train")  # downloads the full dataset
+print(imagenet[0])
+```
+
+But one caveat is that you must have the entire dataset stored on your disk or in memory, which blocks you from accessing datasets bigger than the disk.
+Because it can become inconvenient for big datasets, there exists another type of dataset, the [`IterableDataset`].
+When you have an `IterableDataset`, you can access it using a `for` loop to load the data progressively as you iterate over the dataset.
+This way, only a small fraction of examples is loaded in memory, and you don't write anything on disk.
+
+For example, you can stream the ImageNet-1k dataset without waiting for it to download completely:
```
Comment (polinaeterna) [inline docs/source/about_mapstyle_vs_iterable.mdx:26]:
```suggestion
For example, you can stream the ImageNet-1k dataset without downloading it on disk:
```

File: docs/source/about_mapstyle_vs_iterable.mdx [hunk 35]
```diff
@@ -0,0 +1,219 @@
+# Differences between Dataset and IterableDataset
+
+There are two types of dataset objects, a [`Dataset`] and an [`IterableDataset`].
+Whichever type of dataset you choose to use or create depends on the size of the dataset.
+In general, an [`IterableDataset`] is ideal for big datasets (think hundreds of GBs!) due to its lazy behavior and speed advantages, while a [`Dataset`] is great for everything else.
+This page will compare the differences between a [`Dataset`] and an [`IterableDataset`] to help you pick the right dataset object for you.
+
+## Downloading and streaming
+
+When you have a regular [`Dataset`], you can access it using `my_dataset[0]`. This provides "random access" to the rows.
```
Comment (polinaeterna) [inline docs/source/about_mapstyle_vs_iterable.mdx:10]:
```suggestion
When you have a regular [`Dataset`], you can access it using `my_dataset[0]`. This provides random access to the rows.
```
are quotes needed here?

File: docs/source/process.mdx [hunk 36]
```diff
@@ -92,6 +107,8 @@ There are two options for filtering rows in a dataset: [`~Dataset.select`] and [
 1834.0
 ```
 
+Unless the list of indices to keep is contiguous, those methods also create an indices mapping under the .
```
Comment (polinaeterna) [inline docs/source/process.mdx:110]: ... hood? :)

File: docs/source/use_with_pytorch.mdx [hunk 37]
```diff
@@ -200,7 +200,10 @@ You must use a `BatchSampler` if you want the transform to be given full batches
 
 ### Stream data
 
-Loading a dataset in streaming mode allows one to iterate over the dataset without downloading it on disk.
+Stream a dataset by loading it as an [`IterableDataset`]. This allows you to progressively iterate over a dataset without downloading it on disk as well as any local dataset files you have stored.
```
Comment (polinaeterna) [inline docs/source/use_with_pytorch.mdx:203]:
```suggestion
Stream a dataset by loading it as an [`IterableDataset`]. This allows you to progressively iterate over a remote dataset without downloading it on disk as well as any local dataset files you have stored.
```

not sure I understand the meaning of the second part of second sentence ("as well as any local..."). Is it like "as well as <iterate over> any local files"?

File: src/datasets/arrow_dataset.py [hunk 38]
```diff
@@ -4643,6 +4672,120 @@ def extra_nbytes_visitor(array, feature):
...
+        ...     ids.set_epoch(epoch)  # will use effective_seed = seed + epoch to shuffle the shards and for the shuffle buffer when you start iterating
+        ...     for example in ids:
+        ...         pass
+        ```
+        Feel free to also use [`IterableDataset.set_epoch`] when using a PyTorch DataLoader or in distributed setups.
+        """
+        from .iterable_dataset import IterableDataset
+
+        if num_shards > len(self):
+            raise ValueError(f"Unable to shard a dataset of size {len(self)} into {num_shards} shards.")
```
Comment (polinaeterna) [inline src/datasets/arrow_dataset.py:4774]:
```suggestion
            raise ValueError(f"Unable to shard a dataset of size {len(self)} into {num_shards} shards (the number of shards exceeds the number of samples).")
```
or rows?

**Review by polinaeterna — APPROVED** [review]
I **love** the new doc about Dataset vs IterableDataset, thank you! and I think it's worth a separate page. 
I left just a few comments to the text.

**Review by mariosasko — APPROVED** [review]
Looks good! I guess we can keep the new page. 

Btw, this page boils down to "switch from Dataset to IterableDataset to save time/disk space by avoiding the full rewrite of a dataset" (`map` always does it, `flatten_indices` is good to run after `shuffle` to preserve the iteration speed), so maybe a better title for it would be "Optimize processing" (or "Working with datasets at scale" as I mentioned earlier on Slack)

PS: I think it would be a good idea to add links to the Guide pages for better discoverability and to somewhat "justify their presence in the docs" (from the tutorial/how-to pages to the guides; some guides are not referenced at all) 
cc @stevhliu

File: docs/source/use_with_pytorch.mdx (hunk 37 above)
Comment (lhoestq) [reply docs/source/use_with_pytorch.mdx:203]: yes exactly, let me rephrase

**Comment by lhoestq** [comment]
Took your last comments into account !

> so maybe a better title for it would be "Optimize processing" (or "Working with datasets at scale" as I mentioned earlier on Slack)

I think the content would be slightly different, e.g. focus more on multiprocessing or what data formats to use. This can be a complementary page IMO

> PS: I think it would be a good idea to add links to the Guide pages for better discoverability and to somewhat "justify their presence in the docs" (from the tutorial/how-to pages to the guides; some guides are not referenced at all)

Added a link in the how-to stream page. We may want to include it in the tutorial at one point at well - right now none of the tutorials mention streaming

**Comment by stevhliu** [comment]
> PS: I think it would be a good idea to add links to the Guide pages for better discoverability and to somewhat "justify their presence in the docs" (from the tutorial/how-to pages to the guides; some guides are not referenced at all)

Just merged #5485, which references this new doc! Will look for other pages in the docs where it'd make sense to add them :)
